src/otaclient/_logging.py

```diff
@@ -19,32 +19,181 @@
 import atexit
 import contextlib
 import logging
+import time
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
 from queue import Queue
 from threading import Event, Thread
-from urllib.parse import urljoin
+from urllib.parse import urljoin, urlparse
 
+import grpc
 import requests
+from otaclient_iot_logging_server_pb2.v1 import (
+    otaclient_iot_logging_server_v1_pb2 as log_pb2,
+)
+from otaclient_iot_logging_server_pb2.v1 import (
+    otaclient_iot_logging_server_v1_pb2_grpc as log_v1_grpc,
+)
+from pydantic import AnyHttpUrl
 
 from otaclient.configs.cfg import cfg, ecu_info, proxy_info
 
 
+class LogType(Enum):
+    LOG = "log"
+    METRICS = "metrics"
+
+
+@dataclass
+class QueueData:
+    """Queue data format for logging."""
+
+    log_type: LogType
+    message: str
+
+
+class Transmitter(ABC):
+    def __init__(self, logging_upload_endpoint, ecu_id: str):
+        self.ecu_id = ecu_id
+
```

Comment:
Here I can see that the intention is,  the `ecu_id` will be directly assign to the instance, while the `logging_upload_endpoint` will be parsed by each detailed implementation of `Transmitter`.
But it looks strange to me that, the `__init__` of an ABC base class has partial implementation.

Here I recommend that, we just set the `__init__` as `abstractmethod` also, and this `abstractmethod` is to declare that, the `Transmitter` implementation will take `logging_upload_endpoint` and `ecu_id` args, and completely leave the actual implementation to the detailed subclass.

Reply:
You are right, complete interface will be better than abstract class in this case. Fixed.